pulp_ansible/app/tasks/test_tasks.py

```diff
@@ -0,0 +1,94 @@
+from pulp_ansible.app.tasks.collections import import_collection
+
+
+def create_namespace(base_path, collection_n=20, versions_per_collection=5):
+    namespace_name = "".join([random.choice(string.ascii_lowercase) for i in range(26)])
+    for collection_i in range(collection_n):
+        collection_name = "".join([random.choice(string.ascii_lowercase) for i in range(26)])
+        for version_i in range(versions_per_collection):
+            version = str(version_i + 1) + ".0.0"
+
+            cmd = "ansible-galaxy collection init --force --init-path {} {}.{}".format(
+                base_path, namespace_name, collection_name
+            )
+            subprocess.run(cmd.split())
+
+            path_to_namespace_dir = os.path.join(base_path, namespace_name)
+            path_to_collection_dir = os.path.join(base_path, namespace_name, collection_name)
+            galaxy_yml_path = os.path.join(path_to_collection_dir, "galaxy.yml")
+            with fileinput.FileInput(galaxy_yml_path, inplace=True) as file:
+                for line in file:
+                    print(line.replace("1.0.0", version), end='')
+
+            cmd = "ansible-galaxy collection build --output-path {} {}".format(
+                base_path, path_to_collection_dir
+            )
+            subprocess.run(cmd.split())
+
+            shutil.rmtree(path_to_namespace_dir)  # Remove the collection build dir
+
+
+def import_collection_from_path(path):
+    artifact = Artifact.init_and_validate(path)
+
+    try:
+        artifact.save()
+    except IntegrityError:
+        artifact = Artifact.objects.get(sha256=artifact.sha256)
+
+    import_collection(artifact.pk)
+
+
+def create_repos_with_collections(repo_count, num_repo_versions, collection_percentage):
+    content_pk = [i for i in CollectionVersion.objects.values_list('pk', flat=True)]
+    last_content_index = math.floor(collection_percentage * len(content_pk))
+
+    for repository_i in range(repo_count):
+        repository, created = AnsibleRepository.objects.get_or_create(name=uuid.uuid4())
+
+        for version_i in range(num_repo_versions):
+            random.shuffle(content_pk)
+            with repository.new_version() as new_version:
+                qs = CollectionVersion.objects.filter(pk__in=content_pk[:last_content_index])
+                new_version.add_content(qs)
+
+
+def promote_content(repos_per_task):
+    # import pydevd_pycharm
+    # pydevd_pycharm.settrace('localhost', port=29437, stdoutToServer=True, stderrToServer=True)
+    random_collection = CollectionVersion.objects.order_by('?').first()
+    repos_to_dispatch = []
+    locks = []
+    for repo in AnsibleRepository.objects.all():
+        repos_to_dispatch.append(repo.pk)
+        locks.append(repo)
+        if len(repos_to_dispatch) == repos_per_task:
+            task_args = (random_collection.pk, repos_to_dispatch)
+            enqueue_with_reservation(promote_content, locks, args=task_args)
```

Comment:
promote_content task being recursive reads a little odd to me, but I suppose it is to pass the locked objects down to each sub-promote?

(is this going to create one too level transaction? Not sure where the transaction entry/exit points would be)

Reply:
This was a bug, it's been updated. Thanks for commenting!